openstack/containerorchestration/v1/common/errors.go

```diff
@@ -0,0 +1,118 @@
+package common
+
+import (
+	"encoding/json"
+	"github.com/gophercloud/gophercloud"
+	"strings"
+)
+
+// ErrorResponse represents failed response
+type ErrorResponse struct {
+	gophercloud.ErrUnexpectedResponseCode
+	Errors []Error `json:"errors" required:"true"`
+}
+
+// Error represents a Magnum API error message
+type Error struct {
+	Status    int    `json:"status"`
+	Code      string `json:"code"`
+	Title     string `json:"title"`
+	Detail    string `json:"detail"`
+	RequestID string `json:"request_id,omitempty"`
+}
+
+// Error returns the error message details from the server
+func (e *ErrorResponse) Error() string {
+	errors := []string{}
+	for _, e := range e.Errors {
+		errors = append(errors, e.Detail)
+	}
+	return strings.Join(errors, "\n")
+}
+
+func (e *ErrorResponse) unwrapError(r gophercloud.ErrUnexpectedResponseCode) bool {
```

Comment:
I've been considering this off and on, and I have a suggestion. First, though, a comment that this whole thing reminds me needs to go in the README:
The way errors are handled currently is more of an experiment than a certain library feature; basically, I'm not sure how I feel about the design of it. It exists for 2 reasons:
1. When we get an authentication error, we'd like to be able to retry the operation (or run a user-defined function).
2. To allow overriding the error message returned by OpenStack, since many times the error messages returned are vague and/or unhelpful
The reason I mention that is because I'm not sure we gain anything by implementing all the `Error***` interfaces here.

Now the suggestion: it looks like the error logic here is to unmarshal the `error` into an `Error` if possible, otherwise return original error. Because of that, could we just create an `ExtractErr` method on `DeleteResult` with that logic? Something like the following (untested):

```go
func (h *DeleteResult) ExtractErr() error {
  if h.Err != nil {
    if e, ok := unwrapError(h.Err); ok {
      return e
    }
    return h.Err
  }
  return nil
}
``